File: apps/dominion/pkg/soap/discovery.go
package soap

import (
	"github.com/mattrax/xml"
)

// DiscoverRequest contains the device and user information to help inform the response
type DiscoverRequest struct {
	XMLName xml.Name      `xml:"s:Envelope"`
	Header  RequestHeader `xml:"s:Header"`
	Body    struct {
		EmailAddress       string       `xml:"EmailAddress"`
		RequestVersion     string       `xml:"RequestVersion"`
		DeviceType         string       `xml:"DeviceType"`
		ApplicationVersion string       `xml:"ApplicationVersion"`
		OSEdition          string       `xml:"OSEdition"`
		AuthPolicies       AuthPolicies `xml:"AuthPolicies"`
	} `xml:"s:Body>Discover>request"`
}

// DiscoverResponse contains the enrollment endpoints and authentication type for the device to continue enrollment with
type DiscoverResponse struct {
	XMLName                    xml.Name `xml:"http://schemas.microsoft.com/windows/management/2012/01/enrollment DiscoverResponse"`
	AuthPolicy                 string   `xml:"DiscoverResult>AuthPolicy"`
	EnrollmentVersion          string   `xml:"DiscoverResult>EnrollmentVersion"`
	EnrollmentPolicyServiceURL string   `xml:"DiscoverResult>EnrollmentPolicyServiceUrl"`
	EnrollmentServiceURL       string   `xml:"DiscoverResult>EnrollmentServiceUrl"`
	AuthenticationServiceURL   string   `xml:"DiscoverResult>AuthenticationServiceUrl,omitempty"`
}

// AuthPolicies contains the array of supported AuthPolicies
type AuthPolicies struct {
	AuthPolicies []string `xml:"AuthPolicy"`
}

File: packages/ms-mde/src/discovery.ts
import * as v from "valibot";
import { envelope, withEnvelope } from "./envelope";

// DiscoverRequest contains the device and user information to help inform the response
export const discoveryRequest = envelope(
	"http://schemas.microsoft.com/windows/management/2012/01/enrollment/IDiscoveryService/Discover",
	{
		Discover: v.object({
			"@_xmlns": v.literal(
				"http://schemas.microsoft.com/windows/management/2012/01/enrollment",
			),
			request: v.object({
				"@_xmlns:i": v.literal("http://www.w3.org/2001/XMLSchema-instance"),
				EmailAddress: v.pipe(v.string(), v.email()),
				OSEdition: v.number(),
				RequestVersion: v.number(),
				DeviceType: v.string(),
				ApplicationVersion: v.string(),
				AuthPolicies: v.object({
					AuthPolicy: v.pipe(
						v.any(),
						v.transform((input) => (Array.isArray(input) ? input : [input])),
						v.array(v.picklist(["Federated", "OnPremise", "Certificate"])),
					),
				}),
			}),
		}),
	},
);

type DiscoveryResponseArgs = {
	enrollmentVersion: string;
	enrollmentPolicyServiceUrl: string;
	enrollmentServiceUrl: string;
} & (
	| {
			authPolicy: "Federated";
			authenticationServiceUrl: string;
	  }
	| {
			authPolicy: "OnPremise" | "Certificate";
	  }
);

// Contains the enrollment endpoints and authentication type for the device to continue enrollment with
export const discoverResponse = withEnvelope(
	"http://schemas.microsoft.com/windows/management/2012/01/enrollment/IDiscoveryService/DiscoverResponse",
	(opts: DiscoveryResponseArgs) => ({
		DiscoverResponse: {
			"@_xmlns":
				"http://schemas.microsoft.com/windows/management/2012/01/enrollment",
			DiscoverResult: {
				AuthPolicy: opts.authPolicy,
				EnrollmentVersion: opts.enrollmentVersion,
				EnrollmentPolicyServiceUrl: opts.enrollmentPolicyServiceUrl,
				EnrollmentServiceUrl: opts.enrollmentServiceUrl,
				AuthenticationServiceUrl:
					"authenticationServiceUrl" in opts
						? opts.authenticationServiceUrl
						: undefined,
			},
		},
	}),
);
